src/components/Timer.react.js

```diff
@@ -0,0 +1,322 @@
+     * components in an app.
+     */
+    id: PropTypes.string,
+
+    /**
+     * This component will increment the counter `n_intervals` every
+     * `interval` milliseconds
+     */
+    interval: PropTypes.number,
+
+    /**
+     * If True, the n_interval counter  and the timer no longer updates.  This pauses the timer.
+     */
+    disabled: PropTypes.bool,
+
+    /**
+     * Number of times the interval has passed (read-only)
+     */
+    n_intervals: PropTypes.number,
+
+    /**
+     * Number of times the interval will be fired.
+     * If -1, then the interval has no limit (the default)
+     * and if 0 then the interval stops running.
+     */
+    max_intervals: PropTypes.number,
+
+    /**
+     * When in countdown mode, the timer will count down to zero from the starting `duration` and will show the number
+     *  of milliseconds remaining.
+     *  When in stopwatch mode, the timer will count up from zero and show the number of milliseconds elapsed.
+     *  (read-only)
+     */
+    timer: PropTypes.number,
+
+    /**
+     * The timer will count down to zero in `countdown` mode and count up from zero in `stopwatch` mode
+     */
+    mode: PropTypes.oneOf(['stopwatch', 'countdown']),
+
+    /**
+     * Sets the number of milliseconds the timer will run.  If -1 the timer will not be limited by the duration. (the default)
+     * and if 0 then the timer stops running.
+     */
+    duration: PropTypes.number,
+
+    /**
+     * This will start the timer at the beginning with the given prop settings.
+     */
+    reset: PropTypes.bool,
+
+    /**
+     * A list of the time(s) in milliseconds at which to fire a callback. This can be used to start a task at a given
+     * time rather than using the timer. Since the timer is typically set at a small interval like one second, using
+     * fire_times can reduce the number of times a callback is fired and can increase app performance. The time(s) must be a
+     * multiple of the interval.
+     */
+    fire_times: PropTypes.arrayOf(PropTypes.number),
+
+    /**
+     * This is updated when the timer reaches one of the times in the `fire_times` list. Using `at_fire_time` in a
+     * callback will trigger the callback at the time(s) in `fire_times` (Read only)
+     */
+    at_fire_time: PropTypes.number,
+
+    /**
+     * When True, the timer repeats once the timer has run for the number of milliseconds set in the duration.
+     */
+    rerun: PropTypes.bool,
+
+    /**
+     * Timer messages to be displayed by the component rather than showing the timer. It is a dictionary in the form of:
+     * {integer: string} where integer is the time in milliseconds of when the string message is to be displayed.
+     * Note: timer_format will override messages. For example, {10000 : "updating in 10 seconds"} will display the message
+     * "updating in 10 seconds" once the timer equals 10000.
+     */
+    messages: PropTypes.object,
```

Comment:
This can be `PropTypes.objectOf(PropTypes.string)` as in [Slider](https://github.com/plotly/dash-core-components/blob/50cea817f7da59809b5f2e41f2d0dadbbc5e54e7/src/components/Slider.react.js#L36)

Do these times also need to be multiples of `interval`? And if so, can we relax that too, so the message is displayed when we reach or pass the specified time?

Reply:
For now it needs to be a multiples of the `interval` else the timer will never be equal to the time in the `messages` dict.  Unlike with `dcc.Interval` it's possible to have a lot of functionality clientside with this component so there is no   performance issue with making the `interval` a small number.  Then it's easy to make the times in the `message`, `fire_times` and the `duration` a multiple of the `interval`.

If we leave it as-is, should there be an warning message if the `duration` or `fire_times` or `message` times are not multiples of the `interval`?

Reply:
OK, let's keep the requirement that `messages` keys be multiples of `interval` for now as well.

We could make a console warning when any of these is not a multiple of interval. That's not very prominent, in fact we normally try to avoid giving anyone a reason to look in the console. I'm tempted to say we should actually throw an error. Seems a bit heavy-handed, but things will start to fail in hard-to-debug ways when that happens so maybe it's the best option. For example: if `duration` is not a multiple, the timer will still run for about the right amount of time but `fire_times` and `messages`, even if they ARE multiples, will not respond correctly because the actual `time` values will be `duration - n_intervals * interval` so also not multiples of `interval`.